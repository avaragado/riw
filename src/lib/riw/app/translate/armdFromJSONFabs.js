// @flow

import fs from 'fs';

import compose from 'ramda/src/compose';

const armdFromJSONFabs: MessageDescriptorsFromFile = compose(
    JSON.parse,
    fs.readFileSync,
);

export default armdFromJSONFabs;
