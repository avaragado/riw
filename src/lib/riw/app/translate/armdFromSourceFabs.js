// @flow

import { transformFileSync, type babel } from 'babel-core';
import pluginReactIntl from 'babel-plugin-react-intl';
import pathOr from 'ramda/src/pathOr';
import compose from 'ramda/src/compose';

import type { AbsolutePath } from '../../../../types';
import log from '../../../log';

import type { MessageDescriptorsFromFile } from './extract';


const outputBabelFromFabs = (fabs: AbsolutePath): babel.BabelFileResult => {
    try {
        return transformFileSync(fabs, {
            plugins: [pluginReactIntl],
        });

    } catch (err) {
        log.error(err);
        throw err;
    }
};

const armdFromFabs: MessageDescriptorsFromFile = compose(
    pathOr([], ['metadata', 'react-intl', 'messages']),
    outputBabelFromFabs,
);

export default armdFromFabs;
