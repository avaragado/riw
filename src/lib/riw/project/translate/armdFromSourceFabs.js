// @flow

import { transformFileSync, type babel } from 'babel-core';
import pluginReactIntl from 'babel-plugin-react-intl';
import pathOr from 'ramda/src/pathOr';
import compose from 'ramda/src/compose';

const outputBabelFromFabs = (fabs: AbsolutePath): babel.BabelFileResult => {
    try {
        return transformFileSync(fabs, {
            plugins: [pluginReactIntl],
        });

    } catch (err) {
        console.log(err);
        throw err;
    }
};

const armdFromFabs: MessageDescriptorsFromFile = compose(
    pathOr([], ['metadata', 'react-intl', 'messages']),
    outputBabelFromFabs,
);

export default armdFromFabs;
