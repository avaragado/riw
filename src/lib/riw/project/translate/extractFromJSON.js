// @flow

import fs from 'fs';
import path from 'path';

import glob from 'glob';
import compose from 'ramda/src/compose';

const arfabsFromConfig: FilesFromConfig = config =>
    glob.sync(
        '**/*.json',
        {
            cwd: path.resolve(config.rootDir || '.', config.collateDir),
            absolute: true,
        },
    );

const armdFromFabs: MessageDescriptorsFromFile = compose(
    JSON.parse,
    fs.readFileSync,
);

export {
    arfabsFromConfig,
    armdFromFabs,
};
