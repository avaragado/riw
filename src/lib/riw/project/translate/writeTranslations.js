// @flow

import fs from 'fs';
import path from 'path';

import mkdirp from 'mkdirp';

import log from '../../../log';
import translationsOutputFile from '../../../translationsOutputFile';

type Notifier = (name: string) => (fabs: AbsolutePath) => AbsolutePath;
type Writer = (
    config: RIWConfig,
    notify: Notifier,
    tof: {
        fromLid: (lid: LocaleId) => AbsolutePath,
        forSingleFile: () => AbsolutePath,
    },
    translation: RIWFindTranslationResult,
) => void;

type WriterMap = { [key: OutputMode]: Writer };

const stringify = obj => JSON.stringify(obj, null, 4);

const arlidInclude = (config: RIWConfig): LocaleId[] =>
    config.targetLocales.concat(config.defaultLocale);

const arWriterByMode: WriterMap = {
    'file-per-locale': (config, notify, tof, translation) => {
        arlidInclude(config).forEach((lid) => {
            const json = stringify(translation.locale[lid] || {});
            const fabs = tof.fromLid(lid);

            fs.writeFileSync(fabs, json);
            notify('fileSaved')(fabs);
        });
    },

    'single-file': (config, notify, tof, translation) => {
        const data = arlidInclude(config).reduce(
            (dataAcc, lid) => {
                dataAcc[lid] = translation.locale[lid] || {};
                return dataAcc;
            },
            {},
        );

        const json = stringify(data);
        const fabs = tof.forSingleFile();

        fs.writeFileSync(fabs, json);
        notify('fileSaved')(fabs);
    },

    'no-file': () => {},
};

export default (config: RIWConfig, notify: Notifier) =>
    (translation: RIWFindTranslationResult) => {
        const tof = translationsOutputFile(config);
        const { translationsOutputFile: pathOut } = config;

        if (!tof.hasPlaceholder()) {
            log.error('riw', `Include "${tof.sPlaceholder}" placeholder somewhere in your translationsOutputFile setting.`);
            log.error('riw', `Current translationsOutputFile setting: ${pathOut}`);

            throw new Error(`Configuration error: translationsOutputFile needs "${tof.sPlaceholder}`);
        }

        const dabsOut = path.dirname(tof.forSingleFile());

        try {
            mkdirp.sync(dabsOut);
        } catch (err) {
            log.error('riw', `Unable to create directory ${dabsOut} for riw translations file(s).`);
            throw err;
        }

        arWriterByMode[config.outputMode](config, notify, tof, translation);

        return translation;
    };
