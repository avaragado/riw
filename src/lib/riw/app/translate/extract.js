// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import { arfabsInputJSON, arfabsInputSource } from '../../../config-helper';

import armdFromJSONFabs from './armdFromJSONFabs';
import armdfromSourceFabs from './armdFromSourceFabs';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, file: fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(arfabsInputSource, armdfromSourceFabs);
export const armdExtractJSON = armdExtract(arfabsInputJSON, armdFromJSONFabs);
