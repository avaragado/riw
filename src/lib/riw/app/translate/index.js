// @flow

import compose from 'ramda/src/compose';

import type {
    AbsolutePath,
    MessageDescriptorWithFile,
} from '../../../../types';
import type { Config } from '../../../config';
import type { Notifier } from '../../../notify';
import makeNotifier from '../../../notify';

import db from '../../db';

import type { TranslationLookupResult } from './findTranslations';
import type { DuplicateIdData } from './findDuplicateIds';
import { armdExtractSource, armdExtractJSON } from './extract';
import ardupFind from './findDuplicateIds';
import findTranslations from './findTranslations';
import writeTranslations from './writeTranslations';
import writeTodo from './writeTodo';

export type {
    TranslationLookupResult,
    DuplicateIdData,
};

export type AppTranslateResult = {
    armd: MessageDescriptorWithFile[],
    dups: DuplicateIdData[],
    translation: TranslationLookupResult,
};

export type AppTranslateSpec = {
    on?: {
        start?: () => void,
        startExtract?: (arfabsSource: AbsolutePath[]) => void,
        startExtractFile?: (fabsSource: AbsolutePath) => void,
        endExtractFile?: ({ armd: MessageDescriptorWithFile[], fabs: AbsolutePath }) => void,
        endExtract?: (MessageDescriptorWithFile[]) => void,
        startDupCheck?: (armd: MessageDescriptorWithFile[]) => void,
        endDupCheck?: (dups: DuplicateIdData[]) => void,
        startLookup?: (armd: MessageDescriptorWithFile[]) => void,
        endLookup?: (translation: TranslationLookupResult) => void,
        fileSaved?: (fabs: AbsolutePath) => void,
        end?: (result: AppTranslateResult) => void,
    },
};

type Processor = (config: Config, notify: Notifier) => AppTranslateResult;

const process: Processor = (config, notify) => compose(
    notify('end'),
    armd => ({
        armd,
        dups: ardupFind(notify)(armd),
        translation: compose(
            writeTodo(config, notify),
            writeTranslations(config, notify),
            findTranslations(config, notify, db(config).read()),
        )(armd),
    }),
    {
        source: armdExtractSource(notify),
        json: armdExtractJSON(notify),
    }[config.inputMode],
    notify('start'),
)(config);

export default (config: Config) => (opt: AppTranslateSpec): AppTranslateResult =>
    process(config, makeNotifier(opt.on));
