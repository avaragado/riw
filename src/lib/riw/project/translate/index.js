// @flow

import compose from 'ramda/src/compose';

import makeNotifier from '../../../notify';

import db from '../../db';

import { armdExtractSource, armdExtractJSON } from './extract';
import ardupFind from './findDuplicateIds';
import findTranslations from './findTranslations';
import writeTranslations from './writeTranslations';
import writeTodo from './writeTodo';

const process = (config, notify) => compose(
    notify('end'),
    armd => ({
        armd,
        dups: ardupFind(notify)(armd),
        translation: compose(
            writeTodo(config, notify),
            writeTranslations(config, notify),
            findTranslations(config, notify, db(config).read()),
        )(armd),
    }),
    {
        source: armdExtractSource(notify),
        json: armdExtractJSON(notify),
    }[config.inputMode],
    notify('start'),
)(config);

export default (config: RIWConfig) =>
    (opt: RIWCLIOptProjectTranslate): RIWCLIProjectTranslateResult =>
    process(config, makeNotifier(opt.on));
