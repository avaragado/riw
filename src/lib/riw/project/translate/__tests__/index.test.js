// @flow

import fs from 'fs';

import mock from 'mock-fs';
import outdent from 'outdent';

import translate from '../';
import cfgBase from '../../../__tests__/helpers/dummyConfig';

const stringify = obj => JSON.stringify(obj, null, 4);
const parse = frel => JSON.parse(fs.readFileSync(frel).toString());

const db: RIWDB = {
    version: 1,
    data: {
        'a.1!': {
            '_': {
                'en-up': 'A.1!',
                'en-ne': '!1.a',
            },
        },
        'b.1!': {
            '_': {
                'en-ne': '!1.b',
            },
        },
        'b.2!': {
            'b.2 desc': {
                'en-up': 'B.2!',
            },
        },
    },
};

const cfgOverride: RIWConfig = {
    ...cfgBase,
    rootDir: '.',
    defaultLocale: 'en-us',
    targetLocales: ['en-up', 'en-ne'],
    translationsDatabaseFile: 'fixtures/db.json',
    sourceDirs: ['fixtures/dir/**/*.js'],
    translationsOutputFile: 'fixtures/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'fixtures/locale/todo.json',
};

type Fixture = {
    name: string,
    in: { [key: string]: string },
    configOverride?: RIWConfigSparse,
};

const fixtures: Fixture[] = [
    {
        name: '01',
        in: {}, // empty
    },

    {
        name: '02',
        in: {
            'a.js': 'export default a => a;',
            'b.js': 'export default b => b + 1;',
        },
    },

    {
        name: '03',
        in: {
            'a.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    a1: {
                        id: 'a.1',
                        defaultMessage: 'a.1!',
                    },
                });
            `,
            'aa.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    a1again: {
                        id: 'a.1',
                        defaultMessage: 'a.1 again!',
                    },
                });
            `,
            'b.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    b1: {
                        id: 'b.1',
                        defaultMessage: 'b.1!',
                    },
                    b2: {
                        id: 'b.2',
                        description: 'b.2 desc',
                        defaultMessage: 'b.2!',
                    },
                });
            `,
            'bb.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    b2again: {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                    },
                });
            `,
            'c.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    c1: {
                        id: 'c.1',
                        defaultMessage: 'c.1!',
                    },
                });
            `,
        },
    },

    {
        name: '04',
        in: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: 'fixtures/dir',
            inputMode: 'json',
        },
    },
];

describe('lib/riw/project/translate', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                fixtures: {
                    'dir': fixture.in,
                    'db.json': stringify(db),
                    'locale': {},
                },
            });

            const cfg: RIWConfig = {
                ...cfgOverride,
                ...fixture.configOverride,
            };

            const received = translate(cfg)({});
            const arfrel = fs.readdirSync('fixtures/locale');
            const contents = arfrel.map(frel => parse(`fixtures/locale/${frel}`));

            mock.restore();

            expect(received).toMatchSnapshot();
            expect(arfrel).toMatchSnapshot();
            expect(contents).toMatchSnapshot();
        });
    });
});
