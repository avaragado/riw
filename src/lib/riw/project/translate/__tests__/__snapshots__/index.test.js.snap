// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/project/translate 01 1`] = `
Object {
  "armd": Array [],
  "dups": Array [],
  "translation": Object {
    "armdu": Array [],
    "locale": Object {},
  },
}
`;

exports[`lib/riw/project/translate 01 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/project/translate 01 3`] = `
Array [
  Object {},
  Object {},
  Object {},
  Array [],
]
`;

exports[`lib/riw/project/translate 02 1`] = `
Object {
  "armd": Array [],
  "dups": Array [],
  "translation": Object {
    "armdu": Array [],
    "locale": Object {},
  },
}
`;

exports[`lib/riw/project/translate 02 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/project/translate 02 3`] = `
Array [
  Object {},
  Object {},
  Object {},
  Array [],
]
`;

exports[`lib/riw/project/translate 03 1`] = `
Object {
  "armd": Array [
    Object {
      "defaultMessage": "a.1!",
      "description": undefined,
      "fabs": "/fixtures/dir/a.js",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "description": undefined,
      "fabs": "/fixtures/dir/aa.js",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "b.1!",
      "description": undefined,
      "fabs": "/fixtures/dir/b.js",
      "id": "b.1",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/fixtures/dir/b.js",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.js",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "c.1!",
      "description": undefined,
      "fabs": "/fixtures/dir/c.js",
      "id": "c.1",
    },
  ],
  "dups": Array [
    Object {
      "arfabs": Array [
        "/fixtures/dir/a.js",
        "/fixtures/dir/aa.js",
      ],
      "id": "a.1",
    },
    Object {
      "arfabs": Array [
        "/fixtures/dir/b.js",
        "/fixtures/dir/bb.js",
      ],
      "id": "b.2",
    },
  ],
  "translation": Object {
    "armdu": Array [
      Object {
        "defaultMessage": "a.1 again!",
        "description": undefined,
        "fabs": "/fixtures/dir/aa.js",
        "id": "a.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "a.1 again!",
        "description": undefined,
        "fabs": "/fixtures/dir/aa.js",
        "id": "a.1",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.1!",
        "description": undefined,
        "fabs": "/fixtures/dir/b.js",
        "id": "b.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2!",
        "description": "b.2 desc",
        "fabs": "/fixtures/dir/b.js",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/fixtures/dir/bb.js",
        "id": "b.2",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/fixtures/dir/bb.js",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "c.1!",
        "description": undefined,
        "fabs": "/fixtures/dir/c.js",
        "id": "c.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "c.1!",
        "description": undefined,
        "fabs": "/fixtures/dir/c.js",
        "id": "c.1",
        "locale": "en-ne",
      },
    ],
    "locale": Object {
      "en-ne": Object {
        "a.1": "!1.a",
        "b.1": "!1.b",
      },
      "en-up": Object {
        "a.1": "A.1!",
        "b.2": "B.2!",
      },
      "en-us": Object {
        "a.1": "a.1 again!",
        "b.1": "b.1!",
        "b.2": "b.2 again!",
        "c.1": "c.1!",
      },
    },
  },
}
`;

exports[`lib/riw/project/translate 03 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/project/translate 03 3`] = `
Array [
  Object {
    "a.1": "!1.a",
    "b.1": "!1.b",
  },
  Object {
    "a.1": "A.1!",
    "b.2": "B.2!",
  },
  Object {
    "a.1": "a.1 again!",
    "b.1": "b.1!",
    "b.2": "b.2 again!",
    "c.1": "c.1!",
  },
  Array [
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/fixtures/dir/aa.js",
      "id": "a.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/fixtures/dir/aa.js",
      "id": "a.1",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/fixtures/dir/b.js",
      "id": "b.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/fixtures/dir/b.js",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.js",
      "id": "b.2",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.js",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/fixtures/dir/c.js",
      "id": "c.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/fixtures/dir/c.js",
      "id": "c.1",
      "locale": "en-ne",
    },
  ],
]
`;

exports[`lib/riw/project/translate 04 1`] = `
Object {
  "armd": Array [
    Object {
      "defaultMessage": "a.1!",
      "fabs": "/fixtures/dir/a.json",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/fixtures/dir/aa.json",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/fixtures/dir/b.json",
      "id": "b.1",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/fixtures/dir/b.json",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.json",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/fixtures/dir/c.json",
      "id": "c.1",
    },
  ],
  "dups": Array [
    Object {
      "arfabs": Array [
        "/fixtures/dir/a.json",
        "/fixtures/dir/aa.json",
      ],
      "id": "a.1",
    },
    Object {
      "arfabs": Array [
        "/fixtures/dir/b.json",
        "/fixtures/dir/bb.json",
      ],
      "id": "b.2",
    },
  ],
  "translation": Object {
    "armdu": Array [
      Object {
        "defaultMessage": "a.1 again!",
        "fabs": "/fixtures/dir/aa.json",
        "id": "a.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "a.1 again!",
        "fabs": "/fixtures/dir/aa.json",
        "id": "a.1",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.1!",
        "fabs": "/fixtures/dir/b.json",
        "id": "b.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2!",
        "description": "b.2 desc",
        "fabs": "/fixtures/dir/b.json",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/fixtures/dir/bb.json",
        "id": "b.2",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/fixtures/dir/bb.json",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "c.1!",
        "fabs": "/fixtures/dir/c.json",
        "id": "c.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "c.1!",
        "fabs": "/fixtures/dir/c.json",
        "id": "c.1",
        "locale": "en-ne",
      },
    ],
    "locale": Object {
      "en-ne": Object {
        "a.1": "!1.a",
        "b.1": "!1.b",
      },
      "en-up": Object {
        "a.1": "A.1!",
        "b.2": "B.2!",
      },
      "en-us": Object {
        "a.1": "a.1 again!",
        "b.1": "b.1!",
        "b.2": "b.2 again!",
        "c.1": "c.1!",
      },
    },
  },
}
`;

exports[`lib/riw/project/translate 04 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/project/translate 04 3`] = `
Array [
  Object {
    "a.1": "!1.a",
    "b.1": "!1.b",
  },
  Object {
    "a.1": "A.1!",
    "b.2": "B.2!",
  },
  Object {
    "a.1": "a.1 again!",
    "b.1": "b.1!",
    "b.2": "b.2 again!",
    "c.1": "c.1!",
  },
  Array [
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/fixtures/dir/aa.json",
      "id": "a.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/fixtures/dir/aa.json",
      "id": "a.1",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/fixtures/dir/b.json",
      "id": "b.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/fixtures/dir/b.json",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.json",
      "id": "b.2",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/fixtures/dir/bb.json",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/fixtures/dir/c.json",
      "id": "c.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/fixtures/dir/c.json",
      "id": "c.1",
      "locale": "en-ne",
    },
  ],
]
`;
